File: pkg/server/api/console/v1/types/cookie.go
// Copyright (c) 2026 Probo Inc <hello@getprobo.com>.
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
// REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
// AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
// INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
// LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
// OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
// PERFORMANCE OF THIS SOFTWARE.

package types

import (
	"go.probo.inc/probo/pkg/coredata"
	"go.probo.inc/probo/pkg/gid"
	"go.probo.inc/probo/pkg/page"
)

type (
	CookieOrderBy OrderBy[coredata.CookieOrderField]

	CookieConnection struct {
		TotalCount int
		Edges      []*CookieEdge
		PageInfo   PageInfo

		Resolver any
		ParentID gid.GID
	}
)

func NewCookieConnection(
	p *page.Page[*coredata.Cookie, coredata.CookieOrderField],
	parentType any,
	parentID gid.GID,
) *CookieConnection {
	edges := make([]*CookieEdge, len(p.Data))

	for i := range edges {
		edges[i] = NewCookieEdge(p.Data[i], p.Cursor.OrderBy.Field)
	}

	return &CookieConnection{
		Edges:    edges,
		PageInfo: *NewPageInfo(p),

		Resolver: parentType,
		ParentID: parentID,
	}
}

func NewCookieEdge(c *coredata.Cookie, orderBy coredata.CookieOrderField) *CookieEdge {
	return &CookieEdge{
		Cursor: c.CursorKey(orderBy),
		Node:   NewCookie(c),
	}
}

func NewCookie(c *coredata.Cookie) *Cookie {
	return &Cookie{
		ID: c.ID,
		CookieCategory: &CookieCategory{
			ID: c.CookieCategoryID,
			CookieBanner: &CookieBanner{
				ID: c.CookieBannerID,
			},
		},
		Name:        c.Name,
		Duration:    c.Duration,
		Description: c.Description,
		CreatedAt:   c.CreatedAt,
		UpdatedAt:   c.UpdatedAt,
	}
}

File: packages/cookie-banner/src/cookie.ts
// Copyright (c) 2026 Probo Inc <hello@probo.com>.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import type { ConsentAction } from "./types";

export const COOKIE_NAME = "probo_consent";
const SECONDS_PER_DAY = 86400;

export interface ConsentCookie {
  bid: string;
  v: number;
  vid: string;
  action: ConsentAction;
  data: Record<string, boolean>;
}

export function getConsentCookie(): ConsentCookie | null {
  try {
    const prefix = `${COOKIE_NAME}=`;
    const entry = document.cookie
      .split("; ")
      .find((c) => c.startsWith(prefix));

    if (!entry) {
      return null;
    }

    return JSON.parse(decodeURIComponent(entry.substring(prefix.length)));
  } catch {
    return null;
  }
}

export function setConsentCookie(
  value: ConsentCookie,
  expiryDays: number,
): void {
  const maxAge = expiryDays * SECONDS_PER_DAY;
  const encoded = encodeURIComponent(JSON.stringify(value));

  document.cookie = `${COOKIE_NAME}=${encoded}; path=/; max-age=${maxAge}; SameSite=Lax`;
}

export function clearConsentCookie(): void {
  document.cookie = `${COOKIE_NAME}=; path=/; max-age=0; SameSite=Lax`;
}
